File: Java/69_Sqrt(x).java
class Solution {
    public int mySqrt(int x) {
        if (x == 0 || x == 1) return x;

        long start = 1, end = x, mid = -1;
        while (start <= end){
            mid = start + (end - start)/2;
            if (mid*mid > (long) x) end = mid - 1;
            else if(mid*mid == (long) x) return (int)mid;
            else start = mid + 1; 
        }
        return Math.round((int)end);
    }
}

File: C++/69_Sqrt(x).cpp
class Solution {
public:
    int mySqrt(int x) {
        if (x == 0 || x == 1) return x;

        int start = 1, end = x, mid;
        while (start <= end){
            mid = start + (end - start)/2;
            long long square = static_cast<long long>(mid)*mid;
            if (square > x) end = mid - 1;
            else if(square < x) start = mid + 1;
            else return mid; 
        }
        return static_cast<int>(std::round(end));
    }
};
